src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import order.upper_lower
+
+/-!
+# Young diagrams
+
+A Young diagram is a finite set of up-left justified boxes:
+
+```text
+□□□□□
+□□□
+□□□
+□
+```
+This Young diagram corresponds to the [5, 3, 3, 1] partition of 12.
+
+We represent it as a lower set in `ℕ × ℕ` in the product partial order. We write `(i, j) ∈ μ`
+to say that `(i, j)` (in matrix coordinates) is in the Young diagram `μ`.
+
+## Main functions:
+
+  - `young_diagram` : basic definition
+  - `young_diagram.card` : the cardinality (number of boxes)
+
+  - various constructors:
+      `young_diagram.has_emptyc` : the empty Young diagram is (∅ : young_diagram)
+
+## Notation:
+
+In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). This terminology is used
+below, e.g. in `young_diagram.nw_of`.
+
+## Tags
+
+Young diagram
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+section young_diagram
+
+/-- A Young diagram is a lower set in ℕ × ℕ -/
+@[ext] structure young_diagram :=
+(cells : finset (ℕ × ℕ))
+(is_lower_set : is_lower_set (cells : set (ℕ × ℕ)))
+
+instance young_diagram.set_like : set_like young_diagram (ℕ × ℕ) :=
+{ coe            := λ μ, (μ.cells : set (ℕ × ℕ)),
+  coe_injective' := λ μ ν h, by { rwa [young_diagram.ext_iff, ← finset.coe_inj] } }
+
+@[simp] lemma young_diagram.mem_cells {μ : young_diagram} (c : ℕ × ℕ) :
+  c ∈ μ.cells ↔ c ∈ μ := iff.rfl
+
+/-- In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+    means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). -/
+lemma young_diagram.nw_of (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
```

Comment:
Mathlib naming would suggest
```suggestion
lemma young_diagram.mem_of_le_of_mem (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
```

Reply:
Hmm, I could do this, though it would be nice to keep `nw_of` as a synonym? Directional naming will occur elsewhere in the library too (see e.g. `ssyt.legal` and `ssyt.legal.replace` here: https://github.com/jakelev/lean-rsk/blob/dbd97f8fe9fc2ba13d080d37e298ae87d03ff541/src/ssyt.lean#L162). It’s also helpful language for thinking through proofs.

Reply:
This directional naming convention looks confusing, I must say (many things could be called `up` or `down`!). It will probably be better to use longer and less ambiguous names, even if that's slightly unnatural.

Reply:
Orienting Young diagrams like this is very standard, so it shouldn't lead to confusion.

Jake: I suggest creating a zulip thread to discuss this naming more broadly (the basic question: do we permit naming that uses standard directional conventions?) and then make a poll after a few options surface.

One option: `nw_mem_of_mem`. Another, using a tikzcd like naming: `ul_mem_of_mem` (up left).

Reply:
I think the thing I find weird here is that we have to explain which convention we're using before the name makes sense; but elsewhere we don't need to do that, we're using the mathlib convention.

My leaning would be to put the standard directional interpretation in the docstring, just like we don't give lemmas their textbook names.

A Zulip thread is probably a good idea still.

Reply:
I vote for the mathlib-like name, seems much more self-descriptive.

Reply:
@eric-wieser It's not unheard of to include naming conventions in module docs though, and I don't think we should be designing things for someone who isn't familiar with Young diagrams. You need some language to be able to explain the positional relationships between cells.

The so-called English convention is that the cells of a Young diagram are laid out the way matrices are, and you speak of cells as being "above", "below", "left", and "right" of others, and they are in "rows" and "columns". I'm not so keen on using "nw" here because of the pre-existing language, but still, we should have more sophisticated language than just "le".

Reply:
I agree with Kyle, this notation is standard. Up/down/left/right also fits with referring to rows and columns of the young diagram (which isn’t in this PR but will be in a subsequent one). Also, while the directions may seem interchangeable as far as `young_diagram` is concerned, the moment you start doing Young tableaux or representation theory, the rules for vertical and horizontal placement will be different and important to get right. For example, jeu de taquin (https://en.m.wikipedia.org/wiki/Jeu_de_taquin) which is perhaps the single most important algorithm on Young tableaux.

Reply:
I have renamed it to `young_diagram.up_left_mem` for now.